donkeycar/tests/test_path.py

```diff
@@ -56,3 +56,175 @@ def test_csvpath_load(self):
             self.assertEqual((3, 4), xy[1])
 
 
+class TestCTE(unittest.TestCase):
+
+    def test_nearest_two_pts(self):
+        # path containing [(0,1)..(0,100)]
+        path = []
+        for i in range(100):
+            path.append((0, i+1))
+
+        cte = CTE()
+
+        #
+        # by default it finds the nearest point on the path, then
+        # picks the point before it and after it as the segement.
+        #
+        self.assertEqual(((0, 49), (0, 51)), cte.nearest_two_pts(path, 1, 50))
+
+        #
+        # it should handle wrap around at the begining and end of path
+        #
+        self.assertEqual(((0, 100), (0, 2)), cte.nearest_two_pts(path, 1, 1))
+        self.assertEqual(((0, 99), (0, 1)), cte.nearest_two_pts(path, 1, 100))
+
+    def test_nearest_pt_defaults(self):
+        # path containing [(0,1)..(0,100),(0,1)..(0,100)]
+        path = []
+        for i in range(100):
+            path.append((0, i+1))
+        for i in range(100):
+            path.append((0, i+1))
+
+        cte = CTE()
+
+        #
+        # by default it finds first instance of the nearest point on the path,
+        #
+        self.assertEqual(((0, 1), 0, 1), cte.nearest_pt(path, 1, 1))
+        self.assertEqual(((0, 50), 49, 1), cte.nearest_pt(path, 1, 50))
+        self.assertEqual(((0, 100), 99, 1), cte.nearest_pt(path, 1, 100))
+
+    def test_nearest_pt_from_pt(self):
+        # path containing [(0,1)..(0,100),(0,1)..(0,100)]
+        path = []
+        for i in range(100):
+            path.append((0, i+1))
+        for i in range(100):
```

Comment:
`path += path`?

Reply:
Actually you have this path in each test. You can make this a pytest fixture.

Reply:
It' not worth the effort to make a fixture; my rule is once something is used in 3 places then you have enough info to decide how to generalize.  But even then, this is 5 lines of code so 1) is it worth the effort, what are we gaining, 2) does it couple code in a way that makes it more understandable (the coupling is a logical outcome of the code's structure and relationship) or does it couple code that is otherwise unrelated, in which case the coupling makes the code harder to maintain, not easier.  In this case the fact that the two tests use the same path is incidental.

Reply:
I did the path += path.

Reply:
Yeah I see, but now comes my point above, because you have the same code in 3 places, it's now changed in two places only. You can actually write the whole thing like `path = list((0, i+1) for i in range(100)) * 2` to please the gods of python.